dataplicity/scan_directory.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import print_function
+
+
+import logging
+import os
+import os.path
+import time
+
+try:
+    # Scandir from stdlib in Python3
+    from os import DirEntry, scandir
+except ImportError:
+    # scandir from pypy on Python 2.7
+    from scandir import DirEntry, scandir
+
+from typing import Dict, Iterable, List, Optional, Set, Tuple, Union, TypedDict
+
+FileInfo = Union[str, Tuple[str, int]]
+DirectoryDict = TypedDict(
+    "DirectoryDict", {"files": List[FileInfo], "dirs": List[str]}, total=False
+)
+ScanResult = TypedDict(
+    "ScanResult", {"root": str, "time": float, "directories": Dict[str, DirectoryDict]},
+)
+_ScanStackEntry = Tuple[str, Iterable[DirEntry]]
+
+log = logging.getLogger("agent")
+
+
+class ScanDirectoryError(Exception):
+    """Unable to scan the directory."""
+
+
+def scan_directory(root_path, file_sizes=False, max_depth=10):
```

Comment:
This code uses scandir which is available on Python3.5+ or as a backport.

Scandir makes it quiet efficient, and there is no code-level recursion. It will also handle recursive links which otherwise might turn this in to an infinite loop.